ctapipe/io/eventfilereader.py

```diff
@@ -149,215 +98,202 @@ def check_file_compatibility(file_path):
 
     @property
     @abstractmethod
-    def num_events(self):
+    def camera(self):
         """
-        Abstract property to be defined in child class.
+        Name of the camera contained in the file. Read from the file or
+        inferred from the EventFileReader class if the data format is only
+        used by one camera.
 
-        Obtain the number of events from the file, store it inside
-        self._num_events, and return the value.
+        This property is used to choose the correct algorithms to process the
+        waveforms, as some cameras require different algorithms to others
+        (such as with R1 calibration and charge extraction).
 
         Returns
         -------
-        self._num_events : int
+        camera_name : str
         """
 
     @property
-    @abstractmethod
-    def event_id_list(self):
+    def is_stream(self):
         """
-        Abstract property to be defined in child class.
+        Bool indicating if input is a stream. If it is then `__getitem__` is
+        disabled.
 
-        Obtain the number of events from the file, store it as
-        self._event_id_list, and return the value.
+        TODO: Define a method to detect if it is a stream
 
         Returns
         -------
-        self._event_id_list : list[self._num_events]
+        bool
+            If True, then input is a stream.
         """
+        return False
 
     @abstractmethod
-    def read(self, allowed_tels=None, requested_event=None,
-             use_event_id=False):
+    def _generator(self):
         """
         Abstract method to be defined in child class.
 
-        Read the file using the processes required by the readers file-type.
-
-        Parameters
-        ----------
-        allowed_tels : list[int]
-            select only a subset of telescope, if None, all are read. This can
-            be used for example emulate the final CTA data format, where there
-            would be 1 telescope per file (whereas in current monte-carlo,
-            they are all interleaved into one file)
-        requested_event : int
-            Seek to a paricular event index
-        use_event_id : bool
-            If True ,'requested_event' now seeks for a particular event id
-            instead of index
+        Generator where the filling of the `ctapipe.io.containers` occurs.
```

Comment:
I think we should have a single method that fills a single event container. Not just a method that creates a generator for all events.

Reply:
I tried implementing it that way, and it ended up being very complex (and would require a re-write of pyhessio).  It's not a fundamental problem, but we can maybe leave that to a future refactoring and keep it the way it is for now.